test/jdk/java/net/httpclient/HttpResponseLimitingTest.java

```diff
@@ -0,0 +1,273 @@
+    @ParameterizedTest
+    @MethodSource("sufficientCapacities")
+    void testSuccessOnSufficientCapacity(HttpClient.Version version, boolean secure, long sufficientCapacity) throws Exception {
+        HttpResponse<byte[]> response = requestBytes(version, secure, sufficientCapacity);
+        assertArrayEquals(RESPONSE_BODY, response.body());
+    }
+
+    static Arguments[] sufficientCapacities() {
+        return capacityArgs(Long.MAX_VALUE, RESPONSE_BODY.length);
+    }
+
+    @ParameterizedTest
+    @MethodSource("insufficientCapacities")
+    void testFailureOnInsufficientCapacity(HttpClient.Version version, boolean secure, long insufficientCapacity) {
+        assertThrows(
+                IOException.class,
+                () -> requestBytes(version, secure, insufficientCapacity),
+                "body exceeds capacity: " + RESPONSE_BODY.length);
+    }
+
+    static Arguments[] insufficientCapacities() {
+        return capacityArgs(0, RESPONSE_BODY.length - 1);
+    }
+
+    private static Arguments[] capacityArgs(long... capacities) {
+        return Stream
+                .of(HttpClient.Version.HTTP_1_1, HttpClient.Version.HTTP_2)
+                .flatMap(version -> Stream
+                        .of(true, false)
+                        .flatMap(secure -> Arrays
+                                .stream(capacities)
+                                .mapToObj(capacity -> Arguments.of(version, secure, capacity))))
+                .toArray(Arguments[]::new);
+    }
+
+    private static HttpResponse<byte[]> requestBytes(
+            HttpClient.Version version,
+            boolean secure,
+            long capacity)
+            throws Exception {
+
+        // Create the server and the request URI
+        SSLContext sslContext;
+        HttpServerAdapters.HttpTestServer server;
+        String handlerPath = "/";
+        URI requestUri;
+        if (secure) {
+            sslContext = new SimpleSSLContext().get();
+            server = HttpServerAdapters.HttpTestServer.create(version, sslContext);
+            requestUri = URI.create("https://" + server.serverAuthority() + handlerPath);
+        } else {
+            sslContext = null;
+            server = HttpServerAdapters.HttpTestServer.create(version);
+            requestUri = URI.create("http://" + server.serverAuthority() + handlerPath);
+        }
+
+        // Register the request handler
+        server.addHandler(
+                (exchange) -> {
+                    exchange.sendResponseHeaders(200, RESPONSE_BODY.length);
+                    try (var outputStream = exchange.getResponseBody()) {
+                        outputStream.write(RESPONSE_BODY);
+                    }
+                    exchange.close();
+                },
+                handlerPath);
+
+        // Start the server and the client
+        server.start();
+        try (var client = createClient(sslContext)) {
+
+            // Issue the request
+            var request = HttpRequest
```

Comment:
Are we intentionally not setting the HTTP version on the request or the client?

Reply:
I thought it wasn't necessary. Nevertheless, added in e10771317ef57f80d25be58df77eb5c5b0083b0c.